Observability-heavy Axum webserver boilerplate for both rust & rust+wasm templates (#50)
fission-codes/rust-template: README.md

##

This template repository is [Fission's][fission] opinionated [Rust][rust],
Rust [web framework][axum], and [Rust][rust]+[WebAssembly][wasm] (Wasm) project

> generator, which leverages the [cargo-generate][cargo-generate] tool.

Reviewer: ```suggestion
generator, which uses the [cargo-generate][cargo-generate] tool.
```

A bit less wordy.

These templates provide various features for getting-up and running with Rust or
Rust *and* Wasm, including:

* README standardization, code of conduct, contribuing guidelines, and
  a consistent project layout
* GitHub issue and pull-request templates
* A highly-opinionated [axum][axum] middleware-extensive webserver, focused heavily on
  observability and flexibility

> * Some choice default Rust dependencies (particulary for Wasm) via

Reviewer: ```suggestion
* Some choice default Rust dependencies (particularly for Wasm) including
```
🚨🚨 "via" alert 😛

- [Getting Started](#getting-started)
- [Contributing](#contributing)
- [Template References](#template-references)

> - [Thanks](#license)

Reviewer 2: ```suggestion
- [Thanks](#thanks)
```

Author: derp, gracias.

- [`tracing-subscriber`][tracing-subscriber] for Rust binary applications
  to collect trace data, such as by logging it to standard output, and
  consume messages emitted by log-instrumented libraries and modules.
- An option to generate a highly opiniated [`axum`][axum] web application stack.

#### 🔋 Batteries Included Web Framework

If you choose to run a webserver with [`axum`][axum] (choosing `true` at the
prompt), you'll be given an extensive web framework to work from, heavily
influenced by [*Composing an observable Rust application*][composing-rust],
among other sources.

- [`anyhow`][anyhow] as a ergonomic and idiomatic alternative for handling
  errors in applications.
- [`axum`][axum] as the fundamenal, already-🔋's included web framework that

>   serves as our foundation, which demonstrates high-level features like:

Reviewer: ```suggestion
  serves as our foundation, which includes high-level features like:
```

>   * Routing requests to handlers with a macro free API.

Reviewer: ```suggestion
  * Routing requests to handlers with a macro free API
```

Bullet points don't need trailing punctuation if they are a single sentence. (But keep them to be consistent unless you want to update them all.)

  level=INFO span_name="HTTP request" span=2251799813685249 span_event=close_span timestamp=2023-01-29T15:06:42.192221Z http.method=GET latency_ms=3 http.client_ip=127.0.0.1:59965 http.host=localhost:3000 trace_id=fa9754fa3142db2c100a8c47f6dd391d http.route=/ping
  ```
- [`proptest`][proptest] and [`criterion`][criterion] for generating inputs and
  running benchmarks (*optional*).
- [`reqwest`][reqwest] as the default HTTP client library and
  [`reqwest-middleware`][reqwest-middleware] as a wrapper around `reqwest` for
  client middleware chaining, giving us retries and tracing out of the box.
- [`sysinfo`][sysinfo] for monitoring process information like memory, disk
  usage, etc. This information is automatically represented and tracked as
  [gauges][metrics-gauge] for [Prometheus][prom] scraping/export for example.
- [`tracing`][tracing] for instrumenting Rust programs to collect structured,
  event-based diagnostic information, going beyond just logging-style
  diagnostics.
- [`tracing-subscriber`][tracing-subscriber] for Rust binary applications

>   to collect trace data, such as by logging it to standard output, and
>   consume messages emitted by log-instrumented libraries and modules.

Reviewer: ```suggestion
  to collect trace data, such as by logging it to standard output, and to
  consume messages emitted by log-instrumented libraries and modules.
```

  For `axum` projects, we include a compositon of tracing subscribers from
  smaller units of behavior, called [layers][tracing-layers], for collecting,
  augmenting, and logging (as structured logs) trace data. These layers tap into
  hooks triggered throughout a [span’s lifecycle][tracing-span-life]. You can
  find them [here](./rust/src.axum/tracing_layers).

  Event logs are formatted in [`logfmt`][logfmt], as a series of key/value
  pairs. The implementation of the log generation is inspired by
  [influxdata's (Influx DB's) version][influx-logfmt].
- [`utoipa`][utoipa] for compile-time, auto-generated [OpenAPI][openapi]
  documentation and serving it via [Swagger UI][swagger]. This works

>   as a [Procedural `attribute macro`][proc-attr-macro], e.g.:

Reviewer: ```suggestion
  as a [Procedural `attribute macro`][proc-attr-macro], for example:
```

  ```rust
  #[utoipa::path(
      get,
      path = "/ping",
      responses(
          (status = 200, description = "Ping successful"),
          (status = 500, description = "Ping not successful", body=AppError)
      )
  )]
  pub async fn get() -> AppResult<StatusCode> {
      Ok(StatusCode::OK)
  }
  ```
- [`wiremock-rs`][wiremock-rs] to provide HTTP mocking to perform black-box

>   testing of Rust applications that interact with third-party client APIs. This

Reviewer: ```suggestion
  testing of Rust applications that interact with third-party client APIs. This is
```

- [rust-nix-template][rust-nix-template]
- [wasm-pack-template][wasm-pack-template]

## Thanks

Major shout-outs to the various contributors of this work, including:

- [@bgins](https://github.com/bgins)
- [@BoisterousCoder](https://github.com/BoisterousCoder)
- [@QuinnWilton](https://github.com/QuinnWilton)
- [@expede](https://github.com/expede)
- [@pauljamescleary](https://github.com/pauljamescleary)
- [@drunkirishcoder](https://github.com/drunkirishcoder)

> - [@jaredmarrow](https://github.com/jaredmorrow)

Reviewer 2: ```suggestion
- [@jaredmorrow](https://github.com/jaredmorrow)
```

;)

Author: thanks @jaredmorrow!
